sqlx-sqlite/src/index.rs

```diff
@@ -0,0 +1,651 @@
+/// | 1       | column2     | 0         | 0          | 1000      |               |               | a                | c                |
+/// | 1       | column2     | 1         | 0          | 1000      |               |               | c                | x                |
+/// | 1       | column2     | 2         | 0          | 1000      |               |               | x                | z                |
+/// 
+/// While we use SQLite in this example, the index could be implemented with other databases or system.
+/// SQLite is just a convenient example that is also very similar to other RDBMS systems that you might use.
+#[derive(Debug)]
+pub struct SQLiteIndex {
+    pool: SqlitePool,
+}
+
+impl Display for SQLiteIndex {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        writeln!(f, "SQLiteIndex()")?;
+        Ok(())
+    }
+}
+
+impl SQLiteIndex {
+    pub fn new(pool: SqlitePool) -> Self {
+        Self { pool }
+    }
+
+    /// Return the filenames / row groups that match the filter
+    /// 
+    /// This function pushes down the filter to the index to get a list of row groups that match
+    /// and hence the files that need to be read.
+    /// 
+    /// The filter is pushed down to the index by converting it to a set of SQL expressions
+    /// that can be evaluated by the index.
+    /// 
+    /// The return value is a list of `(file_name, FileScanPlan)` tuples where `FileScanPlan` contains
+    /// the file metadata and the row groups that need to be scanned.
+    pub async fn get_files(&self, filter: Option<Expr>) -> Result<Vec<(String, FileScanPlan)>> {
+        let (sql, values) = Query::select()
+            .columns(vec![
+                FileStatistics::FileName,
+                FileStatistics::FileSizeBytes,
+                FileStatistics::RowGroupCount,
+            ])
+            .column(ColumnStatistics::RowGroup)
+            .distinct() // could be distinct_on(vec![ColumnStatistics::FileId, ColumnStatistics::RowGroup]) if the backing store supports it
+            .from(FileStatistics::Table)
+            .inner_join(
+                ColumnStatistics::Table,
+                SeaQExpr::col((FileStatistics::Table, FileStatistics::FileId)).equals((ColumnStatistics::Table, ColumnStatistics::FileId)),
+            )
+            .and_where_option(filter.and_then(|f| push_down_filter(&f)))
+            .build_sqlx(SqliteQueryBuilder);
+
+        // TODO: we could aggregate the row groups into an array in the query to transmit less data over the wire
+        // (and maybe avoid the join), leaving that as a TODO since it introduces more complexity and coupling to the index's backing store
+        // Result is in the form of (file_name, file_size, row_group_count, row_group_to_scan)
+        let row_groups: Vec<(String, i64, i64, i64)> = sqlx::query_as_with(&sql, values)
```

Comment:
FWIW you could probably use sqlite to do the aggregations

Like add a `SELECT DISTINCT ...` to the query so you wouldn't have to handle duplicates in the Rust code (handling duplicates is fine, I am just pointing it out

Reply:
There's already a `distinct` in there (currently line 91).
I do think these queries could be a bit fancier, e.g. I'd like it to just return `file_name string, file_size int, row_groups int[]` but I'm not sure that's possible with sqlite, I'm most familiar with postgres. But I do plan on giving it another shot before merging this.

Reply:
Cool -- sorry I didn't see that. Nice@